sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/MessageFluxRequestDrivenCreditFlowIsolatedTest.java

```diff
@@ -0,0 +1,303 @@
+
+        final AtomicInteger flowCalls = new AtomicInteger();
+        final AtomicLong firstFlow = new AtomicLong();
+        final AtomicLong secondFlow = new AtomicLong();
+        doAnswer(invocation -> {
+            final Object[] args = invocation.getArguments();
+            @SuppressWarnings("unchecked")
+            final Supplier<Long> creditSupplier = (Supplier<Long>) args[0];
+            Assertions.assertNotNull(creditSupplier);
+
+            final int calls = flowCalls.incrementAndGet();
+            if (calls == 1) {
+                firstFlow.set(creditSupplier.get());
+            } else if (calls == 2) {
+                secondFlow.set(creditSupplier.get());
+            }
+            return null;
+        }).when(receiver).addCredit(any());
+
+        try (VirtualTimeStepVerifier verifier = new VirtualTimeStepVerifier()) {
+            verifier.create(() -> messageFlux)
+                .then(() -> upstream.next(receiver))
+                .thenRequest(10)    // 10  -  0  + 100  = 110 [accumulatedRequest_110 >= 100]
+                .thenAwait()
+                .thenRequest(20)    // 30  - 110 + 100  = 20  [accumulatedRequest_20  <  100]
+                .thenRequest(20)    // 50  - 130 + 100  = 20  [accumulatedRequest_40  <  100]
+                .thenRequest(20)    // 70  - 150 + 100  = 20  [accumulatedRequest_60  <  100]
+                .thenRequest(20)    // 90  - 170 + 100  = 20  [accumulatedRequest_80  <  100]
+                .thenRequest(20)    // 110 - 190 + 100 =  20  [accumulatedRequest_100 >= 100]
+                .then(() -> upstream.complete())
+                .verifyComplete();
+        }
+
+        Assertions.assertEquals(2, flowCalls.get());
+        Assertions.assertEquals(prefetch + 10, firstFlow.get());
+        Assertions.assertEquals(prefetch, secondFlow.get());
+        verify(receiver).closeAsync();
+        upstream.assertCancelled();
+    }
+
+    @Test
+    @Execution(ExecutionMode.SAME_THREAD)
+    public void shouldSendFlowWhenRequestAccumulatedGreaterThanPrefetch() {
+        final int prefetch = 100;
+        final TestPublisher<ReactorReceiver> upstream = TestPublisher.create();
+        final MessageFlux messageFlux = new MessageFlux(upstream.flux(), prefetch, CreditFlowMode.RequestDriven, retryPolicy);
+
+        final ReactorReceiver receiver = mock(ReactorReceiver.class);
+        when(receiver.receive()).thenReturn(Flux.never());
+        when(receiver.getEndpointStates()).thenReturn(activeThenNeverTerminate());
+        when(receiver.closeAsync()).thenReturn(Mono.empty());
+
+        final AtomicInteger flowCalls = new AtomicInteger();
+        final AtomicLong firstFlow = new AtomicLong();
+        final AtomicLong secondFlow = new AtomicLong();
+        doAnswer(invocation -> {
+            final Object[] args = invocation.getArguments();
+            @SuppressWarnings("unchecked")
+            final Supplier<Long> creditSupplier = (Supplier<Long>) args[0];
```

Comment:
I was wondering if, instead of keeping track of the invocations like this, if we could use ArgumentCaptor and verify it?

Reply:
Not sure how/if that’s possible, because it’s not sufficient to just capture the `Supplier<Long>` references, but we need to access the actual value that `Supplier<Long>.get()` produces as the strategy executes. Though there are multiple `Supplier<Long>` references, all of them are backed by a single `requestAccumulated` Atomic variable, which self clears once current value is accessed through `Supplier<Long>.get()`.